### DISCUSSION TITLE
Savin Deformation Gradient in CSV file

### DISCUSSION THREAD

**Discussion by mvkpenn22** [idaholab/moose#26494] [Q&A Modules: Solid mechanics] [answered]
Hi,

I am trying to save the deformation gradient value for my load case as follows, but I am getting an error:
[AuxKernels]
  [deformation_gradient_xx]
    type = RankTwoAux
    rank_two_tensor = deformation_gradient
    variable = deformation_gradient_xx
    index_i = 0
    index_j = 0
  []
  [deformation_gradient_xy]
    type = RankTwoAux
    rank_two_tensor = deformation_gradient
    variable = deformation_gradient_xy
    index_i = 0
    index_j = 1
  []
  [deformation_gradient_yx]
    type = RankTwoAux
    rank_two_tensor = deformation_gradient
    variable = deformation_gradient_yx
    index_i = 1
    index_j = 0
  []
  [deformation_gradient_yy]
    type = RankTwoAux
    rank_two_tensor = deformation_gradient
    variable = deformation_gradient_yy
    index_i = 1
    index_j = 1
  []

*** ERROR ***
Variable 'deformation_gradient_xx' does not exist in this system[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 10

[31m
*** ERROR ***
Variable 'deformation_gradient_xx' does not exist in this system[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 15

Is this the right way to store the values?

**Comment by GiudGiud** [comment]
Hello

Is deformation gradient a variable?

If you use Debug/show_material_props does it show up as a material properties?
If so there are auxkernels to output those. 

If not can you compute it from the displacements? I know gradients are not well defined at nodes but maybe a projection would be enough?

Guillaume

**Comment by mvkpenn22** [comment]
Hi @GiudGiud 

Thanks for the suggestion, I defined in the kernel to output these and its material properties are defined as deformation_gradient

{{"strain", "total_strain"},
     {"mechanical_strain", "mechanical_strain"},
     {"stress", "stress"},
     {"cauchy_stress", "cauchy_stress"},
     {"deformation_gradient", "deformation_gradient"},
     {"pk1_stress", "pk1_stress"},
     {"pk2_stress", "pk2_stress"},
     {"small_stress", "small_stress"},
     {"elastic_strain", "elastic_strain"},
     {"plastic_strain", "plastic_strain"},
     {"creep_strain", "creep_strain"},
     {"creep_stress", "creep_stress"}}



[AuxVariables]
  [deformation_gradient_xx]
    order = CONSTANT
    family = MONOMIAL
  []
  [deformation_gradient_yy]
    order = CONSTANT
    family = MONOMIAL
  []
  [deformation_gradient_xy]
    order = CONSTANT
    family = MONOMIAL
  []
  [deformation_gradient_yx]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
[deformation_gradient_xx]
type = RankTwoAux
rank_two_tensor = deformation_gradient
variable = deformation_gradient_xx
index_i = 0
index_j = 0
[]
[deformation_gradient_xy]
type = RankTwoAux
rank_two_tensor = deformation_gradient
variable = deformation_gradient_xy
index_i = 0
index_j = 1
[]
[deformation_gradient_yx]
type = RankTwoAux
rank_two_tensor = deformation_gradient
variable = deformation_gradient_yx
index_i = 1
index_j = 0
[]
[deformation_gradient_yy]
type = RankTwoAux
rank_two_tensor = deformation_gradient
variable = deformation_gradient_yy
index_i = 1
index_j = 1
[]


[Outputs]
  [./csv]
    type = CSV
  [../]
  exodus = true
  file_base = grav02b
[]



But when I run the output file in ParaView, I can only see deformation_gradient_x and deformation_gradient_y not the deformation_gradient_xx, deformation_gradient_yy, deformation_gradient_xy, deformation_gradient_yx

**Reply by GiudGiud** [reply]
Oh that a paraview playing tricks. 

Each of the variables you see is actually a vector with 3 components. Just one more selection menu to get the component

**Reply by mvkpenn22** [reply, marked as answer]
Thank you so much,

It worked
